lib/rb-inotify/notifier.rb

```diff
@@ -51,9 +94,13 @@ def self.supports_ruby_io?
     # @return [Notifier]
     # @raise [SystemCallError] if inotify failed to initialize for some reason
     def initialize
-      @fd = Native.inotify_init
+      fd = Native.inotify_init
       @watchers = {}
-      return unless @fd < 0
+      unless fd < 0
+        @handle = Handle.new(fd)
+        ObjectSpace.define_finalizer(self, @handle.method(:close))
```

Comment:
Ah, this had me very confused, for way longer than it should've. The problem in your first example (and thus correspondingly, my code) is not that the object is retained / the finalizer isn't called: it's that the finalizer gets called *with a parameter*, which the method is not expecting... and finalizers swallow exceptions, without so much as a warning.

As I don't really want to add a useless argument to `Handle#close`, I'll still solve it by introducing a proc. (We can use an instance method, because it's on a different object to the one being finalized -- one we specifically do want the finalizer to retain.)